android/libraries/rib-base/src/main/kotlin/com/uber/rib/core/WorkerScopeProvider.kt

```diff
@@ -20,20 +20,30 @@ import com.uber.rib.core.lifecycle.WorkerEvent
 import io.reactivex.CompletableSource
 import io.reactivex.Observable
 import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.collect
-import kotlinx.coroutines.flow.take
-import kotlinx.coroutines.rx2.asFlow
-import kotlinx.coroutines.rx2.rxCompletable
 
 /** [ScopeProvider] for [Worker] instances.  */
-open class WorkerScopeProvider internal constructor(
-  private val workerLifecycle: Flow<WorkerEvent>,
-) : ScopeProvider {
-  internal constructor(workerLifecycleObservable: Observable<WorkerEvent>) : this(workerLifecycleObservable.asFlow())
+public open class WorkerScopeProvider : ScopeProvider {
+  /*
+   * WorkerScopeProvider may be instantiated with a Flow or Observable of WorkerEvent.
+   * We don't convert from one to another for `requestScope`. Instead, we have alternative implementations for each.
+   * This is done to keep Rx <-> Coroutines interop to a minimum.
+   */
+  private val workerLifecycleFlow: Flow<WorkerEvent>?
+  private val workerLifecycleObservable: Observable<WorkerEvent>?
 
-  override fun requestScope(): CompletableSource {
-    return rxCompletable(RibDispatchers.Unconfined) {
-      workerLifecycle.take(2).collect()
-    }
+  internal constructor(workerLifecycleFlow: Flow<WorkerEvent>) {
+    this.workerLifecycleFlow = workerLifecycleFlow
+    this.workerLifecycleObservable = null
+  }
+
+  internal constructor(workerLifecycleObservable: Observable<WorkerEvent>) {
+    this.workerLifecycleFlow = null
+    this.workerLifecycleObservable = workerLifecycleObservable
+  }
+
+  override fun requestScope(): CompletableSource = when {
+    workerLifecycleFlow != null -> workerLifecycleFlow.asScopeCompletable(WorkerEvent.STOP)
+    workerLifecycleObservable != null -> workerLifecycleObservable.skip(1).firstElement().ignoreElement()
+    else -> error("WorkerScopeProvider must have either a lifecycle Flow or lifecycle Observable.")
```

Comment:
Why we need `public` modifier ? And what about this style instead 
```suggestion
 /*
   * WorkerScopeProvider may be instantiated with a Flow or Observable of WorkerEvent.
   * We don't convert from one to another for `requestScope`. Instead, we have alternative implementations for each.
   * This is done to keep Rx <-> Coroutines interop to a minimum.
   */
sealed class WorkerScopeProvider : ScopeProvider {

internal class FlowBased(private val workerLifecycleFlow: Flow<WorkerEvent>) : WorkerScopeProvider() {
  override fun requestScope(): CompletableSource = 
    workerLifecycleFlow.asScopeCompletable(WorkerEvent.STOP)
  }

internal class ObservableBased(private val workerLifecycleObservable: Observable<WorkerEvent>) : WorkerScopeProvider() {
  override fun requestScope(): CompletableSource = 
    workerLifecycleObservable.skip(1).firstElement().ignoreElement()
  }
}
```

Reply:
> why we need public
We are going to make all libs strictly use explicit API, there's an issue opened for that (#545) and a PR (#548)

> what about this style instead
Good call, I actually thought of using a sealed class/interface when coding this. I'll change it.

Reply:
Update: I did not make it sealed because it would break binary compatibility (e.g. mocks on `WorkerScopeProvider` would not work anymore), but nevertheless I made `WorkerScopeProvider` an interface and separated the implementation into two classes.

Reply:
Update #2: Needed to change things a bit to maintain source compatibility. We need to still support `new WorkerScopeProvider(observable)` constructor. Final form is this:
```kotlin
public class WorkerScopeProvider private constructor(delegate: ScopeProvider): ScopeProvider by delegate {
  constructor(lifecycle: Flow<WorkerEvent>) : this(FlowScopeProvider(lifecycle))
  constructor(lifecycle: Observable<WorkerEvent>) : this(ObservableScopeProvider(lifecycle))
}

private class FlowScopeProvider(private val lifecycle: Flow<WorkerEvent>) : ScopeProvider {
  override fun requestScope(): CompletableSource = lifecycle.asScopeCompletable(WorkerEvent.STOP)
}

private class ObservableScopeProvider(private val lifecycle: Observable<WorkerEvent>) : ScopeProvider {
  override fun requestScope(): CompletableSource = lifecycle.skip(1).firstElement().ignoreElement()
}
```